File: canonicalizer-go/internal/canonical/poseidon.go
// Package canonical 用 Go 实现 DDTM-CANONICAL-V1 确定性数据层：IEEE-754→Q16.16
// 量化、548 字节行编解码、BN254 Poseidon2 width-4 置换（常量钉定为
// gnark-crypto v0.20.1）、域标签 sponge H_P，以及 Merkle 树。
//
// 置换是从零编写的（独立的轮结构），但域运算使用 gnark-crypto 的 fr.Element——
// BN254 标量域的原生类型——因此既快又能证明与 gnark-crypto 完全一致。打包的
// KAT 向量在加载时重新推导，任一不匹配即中止。
package canonical

import (
	"encoding/json"
	"fmt"
	"math/big"
	"os"
	"path/filepath"

	"github.com/consensys/gnark-crypto/ecc/bn254/fr"
)

// poseidonParams 对应 specs/poseidon2-bn254-v1.json。
type poseidonParams struct {
	Modulus       string     `json:"modulus"`
	Width         int        `json:"width"`
	FullRounds    int        `json:"full_rounds"`
	PartialRounds int        `json:"partial_rounds"`
	DiagM1        []string   `json:"diag_m1"`
	RoundKeys     [][]string `json:"round_keys"`
	KAT           []struct {
		Input  []string `json:"input"`
		Output []string `json:"output"`
	} `json:"kat"`
}

// Poseidon 持有钉定的 BN254 Poseidon2 width-4 实例。
type Poseidon struct {
	P             *big.Int // 模数，仅用于参考 / 输出格式化
	Diag          [4]fr.Element
	RoundKeys     [][4]fr.Element
	Width         int
	FullRounds    int
	PartialRounds int
	halfFull      int
}

var sharedPoseidon *Poseidon

// LoadPoseidon 读取钉定的参数文件，并用 KAT 自检。
func LoadPoseidon(specPath string) (*Poseidon, error) {
	if specPath == "" {
		specPath = findSpec("poseidon2-bn254-v1.json")
	}
	raw, err := os.ReadFile(specPath)
	if err != nil {
		return nil, fmt.Errorf("read poseidon params: %w", err)
	}
	var p poseidonParams
	if err := json.Unmarshal(raw, &p); err != nil {
		return nil, fmt.Errorf("parse poseidon params: %w", err)
	}

	inst := &Poseidon{
		P:             mustHex(p.Modulus),
		Width:         p.Width,
		FullRounds:    p.FullRounds,
		PartialRounds: p.PartialRounds,
		halfFull:      p.FullRounds / 2,
	}
	for i, h := range p.DiagM1 {
		inst.Diag[i].SetBigInt(mustHex(h))
	}
	inst.RoundKeys = make([][4]fr.Element, len(p.RoundKeys))
	for i, rnd := range p.RoundKeys {
		for j, h := range rnd {
			inst.RoundKeys[i][j].SetBigInt(mustHex(h))
		}
	}

	// 自检：我们的置换必须重现每一个 gnark-crypto KAT。
	for _, kat := range p.KAT {
		var in [4]fr.Element
		for i, h := range kat.Input {
			in[i].SetBigInt(mustHex(h))
		}
		got := inst.Permute(in)
		for i := range got {
			expBI := mustHex(kat.Output[i])
			gotBI := got[i].BigInt(new(big.Int))
			if gotBI.Cmp(expBI) != 0 {
				return nil, fmt.Errorf("poseidon KAT mismatch on input %v", kat.Input)
			}
		}
	}
	sharedPoseidon = inst
	return inst, nil
}

// SharedPoseidon 返回最近一次加载的实例。
func SharedPoseidon() *Poseidon { return sharedPoseidon }

// Permute 施加 width-4 Poseidon2 置换。与 gnark-crypto 完全一致（由 KAT 自检验证）。
func (p *Poseidon) Permute(state [4]fr.Element) [4]fr.Element {
	s := state

	external := func() {
		var t0, t1, t2, t3, t4, t5, tmp fr.Element
		t0.Add(&s[0], &s[1])
		t1.Add(&s[2], &s[3])
		tmp.Double(&s[1]) // 2*s1
		t2.Add(&t1, &tmp) // t1 + 2*s1
		tmp.Double(&s[3]) // 2*s3
		t3.Add(&t0, &tmp) // t0 + 2*s3
		tmp.Double(&t1).Double(&tmp) // 4*t1
		t4.Add(&tmp, &t3)            // 4*t1 + t3
		tmp.Double(&t0).Double(&tmp) // 4*t0
		t5.Add(&tmp, &t2)            // 4*t0 + t2
		s[0].Add(&t3, &t5)
		s[1].Set(&t5)
		s[2].Add(&t2, &t4)
		s[3].Set(&t4)
	}
	sbox := func(i int) {
		var x2, x4 fr.Element
		x2.Square(&s[i])
		x4.Square(&x2)
		s[i].Mul(&x4, &s[i]) // x^5
	}

	external()

	hf := p.halfFull
	// 前半 full 轮
	for i := 0; i < hf; i++ {
		rk := &p.RoundKeys[i]
		for j := 0; j < 4; j++ {
			s[j].Add(&s[j], &rk[j])
		}
		for j := 0; j < 4; j++ {
			sbox(j)
		}
		external()
	}
	// partial 轮
	for i := hf; i < hf+p.PartialRounds; i++ {
		s[0].Add(&s[0], &p.RoundKeys[i][0])
		sbox(0)
		var tot fr.Element
		tot.Add(&s[0], &s[1]).Add(&tot, &s[2]).Add(&tot, &s[3])
		for j := 0; j < 4; j++ {
			var term fr.Element
			term.Mul(&s[j], &p.Diag[j])
			s[j].Add(&term, &tot)
		}
	}
	// 后半 full 轮
	for i := hf + p.PartialRounds; i < p.FullRounds+p.PartialRounds; i++ {
		rk := &p.RoundKeys[i]
		for j := 0; j < 4; j++ {
			s[j].Add(&s[j], &rk[j])
		}
		for j := 0; j < 4; j++ {
			sbox(j)
		}
		external()
	}
	return s
}

// HashPoseidon 即 H_P(tag, elements)：在 width-4 置换之上用 rate-3 sponge 吸收
// [tag, arity, elements...] 并挤压出第 0 个元素。
func (p *Poseidon) HashPoseidon(tag fr.Element, elements []fr.Element) fr.Element {
	msg := make([]fr.Element, 0, 2+len(elements))
	var arity fr.Element
	arity.SetUint64(uint64(len(elements)))
	msg = append(msg, tag, arity)
	msg = append(msg, elements...)

	var state [4]fr.Element // 全零状态
	for idx := 0; idx < len(msg); idx += 3 {
		end := idx + 3
		if end > len(msg) {
			end = len(msg)
		}
		for j := idx; j < end; j++ {
			state[j-idx].Add(&state[j-idx], &msg[j])
		}
		state = p.Permute(state)
	}
	return state[0]
}

func mustHex(s string) *big.Int {
	v, ok := new(big.Int).SetString(s, 0)
	if !ok {
		panic("bad hex: " + s)
	}
	return v
}

// findSpec 从当前目录向上查找 specs/ 文件。
func findSpec(name string) string {
	dir, _ := os.Getwd()
	for i := 0; i < 8; i++ {
		candidate := filepath.Join(dir, "specs", name)
		if _, err := os.Stat(candidate); err == nil {
			return candidate
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	return filepath.Join("specs", name)
}

File: experiments/g1/poseidon.py
"""Poseidon2 BN254 (width 4) permutation and the ``H_P`` domain-tagged sponge.

The constants are the *audited, hardcoded* gnark-crypto v0.20.1 BN254 width-4
constants, exported to ``specs/poseidon2-bn254-v1.json``. The permutation is
reimplemented here in pure integer arithmetic so that Python produces the
exact same field elements as gnark-crypto (and thus gnark). On import the
implementation re-derives the bundled KAT vectors and aborts if any mismatch
is detected — that is the cross-language correctness contract.

``H_P(tag, elements)`` absorbs ``[tag, arity, *elements]`` through a rate-3
sponge over the width-4 permutation (zero initial state, final block
zero-padded) and squeezes element 0. The explicit ``arity`` field prevents
length-extension ambiguity between inputs of different lengths.
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import Sequence

try:
    import gmpy2  # type: ignore

    _HAS_GMPY2 = True
except ImportError:  # pragma: no cover - gmpy2 is a hard runtime recommendation
    _HAS_GMPY2 = False

REPO_ROOT = Path(__file__).resolve().parents[2]
PARAMS_PATH = REPO_ROOT / "specs" / "poseidon2-bn254-v1.json"

_raw = json.loads(PARAMS_PATH.read_text(encoding="utf-8"))
MODULUS = int(_raw["modulus"], 16)
WIDTH = _raw["width"]
FULL_ROUNDS = _raw["full_rounds"]
PARTIAL_ROUNDS = _raw["partial_rounds"]
_HALF_FULL = FULL_ROUNDS // 2
# Diagonal of the internal matrix (mu_i - 1).
_DIAG = tuple(int(x, 16) for x in _raw["diag_m1"])
# Round keys; round i has len(RK[i]) == WIDTH for the hardcoded width-4 set
# (partial rounds store [key, 0, 0, 0], so adding all four is a no-op on the
# trailing zeros — matching gnark-crypto's addRoundKeyInPlace exactly).
_RK = tuple(tuple(int(x, 16) for x in rnd) for rnd in _raw["round_keys"])


def _to_mpz(value: int):
    return gmpy2.mpz(value) if _HAS_GMPY2 else value


P = _to_mpz(MODULUS)
_DIAG_Z = tuple(_to_mpz(v) for v in _DIAG)
_RK_Z = tuple(tuple(_to_mpz(v) for v in rnd) for rnd in _RK)


def _sbox(x):
    """x^5 mod p (square, square, multiply)."""
    return (x * x * x * x * x) % P  # 4 muls; pow(x,5,P) is equivalent.


def _mat_external(s):
    """External matrix for width 4: the M4 circulant (no cross-block sum)."""
    t0 = (s[0] + s[1]) % P
    t1 = (s[2] + s[3]) % P
    t2 = (2 * s[1] + t1) % P
    t3 = (2 * s[3] + t0) % P
    t4 = (4 * t1 + t3) % P
    t5 = (4 * t0 + t2) % P
    return [(t3 + t5) % P, t5, (t2 + t4) % P, t4]


def _mat_internal(s):
    """Internal matrix: input[i] * DiagM1[i] + sum(input)."""
    tot = (s[0] + s[1] + s[2] + s[3]) % P
    return [(s[i] * _DIAG_Z[i] + tot) % P for i in range(WIDTH)]


def _add_round_key(round_index, s):
    rk = _RK_Z[round_index]
    return [(s[i] + rk[i]) % P for i in range(len(rk))]


def permute(state: Sequence[int]) -> list[int]:
    """Apply the Poseidon2 width-4 permutation to a 4-element state.

    Hot path: the round structure is inlined, mat-ops are written out by hand,
    reductions are deferred where safe, and constants are rebound as locals.
    Arithmetic mirrors gnark-crypto exactly (KAT-verified at import).
    """
    p = P
    rk = _RK_Z
    diag = _DIAG_Z
    s0 = _to_mpz(state[0]) % p
    s1 = _to_mpz(state[1]) % p
    s2 = _to_mpz(state[2]) % p
    s3 = _to_mpz(state[3]) % p

    def external(s0, s1, s2, s3):
        t0 = s0 + s1
        t1 = s2 + s3
        t2 = t1 + (s1 << 1)
        t3 = t0 + (s3 << 1)
        t4 = (t1 << 2) + t3
        t5 = (t0 << 2) + t2
        return (t3 + t5) % p, t5 % p, (t2 + t4) % p, t4 % p

    # Initial external matrix.
    s0, s1, s2, s3 = external(s0, s1, s2, s3)

    hf = _HALF_FULL
    end_partial = hf + PARTIAL_ROUNDS
    total = FULL_ROUNDS + PARTIAL_ROUNDS

    # First half of full rounds: addRoundKey + sBox(all) + external.
    for i in range(hf):
        r = rk[i]
        s0 = (s0 + r[0]) % p
        s1 = (s1 + r[1]) % p
        s2 = (s2 + r[2]) % p
        s3 = (s3 + r[3]) % p
        # x^5 = x2 * x4 ; three muls, reduced.
        a = s0 * s0 % p
        s0 = a * a % p * s0 % p
        a = s1 * s1 % p
        s1 = a * a % p * s1 % p
        a = s2 * s2 % p
        s2 = a * a % p * s2 % p
        a = s3 * s3 % p
        s3 = a * a % p * s3 % p
        s0, s1, s2, s3 = external(s0, s1, s2, s3)

    # Partial rounds: addRoundKey(0) + sBox(0) + internal.
    for i in range(hf, end_partial):
        s0 = (s0 + rk[i][0]) % p
        a = s0 * s0 % p
        s0 = a * a % p * s0 % p
        tot = (s0 + s1 + s2 + s3) % p
        s0 = (s0 * diag[0] + tot) % p
        s1 = (s1 * diag[1] + tot) % p
        s2 = (s2 * diag[2] + tot) % p
        s3 = (s3 * diag[3] + tot) % p

    # Second half of full rounds.
    for i in range(end_partial, total):
        r = rk[i]
        s0 = (s0 + r[0]) % p
        s1 = (s1 + r[1]) % p
        s2 = (s2 + r[2]) % p
        s3 = (s3 + r[3]) % p
        a = s0 * s0 % p
        s0 = a * a % p * s0 % p
        a = s1 * s1 % p
        s1 = a * a % p * s1 % p
        a = s2 * s2 % p
        s2 = a * a % p * s2 % p
        a = s3 * s3 % p
        s3 = a * a % p * s3 % p
        s0, s1, s2, s3 = external(s0, s1, s2, s3)

    return [int(s0), int(s1), int(s2), int(s3)]


def hash_poseidon(tag: int, elements: Sequence[int]) -> int:
    """``H_P(tag, elements)`` — the domain-tagged, arity-separated sponge."""
    # Message = [tag, arity, *elements]; the arity field records len(elements).
    msg = [tag % P, _to_mpz(len(elements))]
    msg.extend(_to_mpz(e) % P for e in elements)

    state = [_to_mpz(0)] * WIDTH
    idx = 0
    n = len(msg)
    while idx < n:
        block = msg[idx : idx + 3]
        for j, value in enumerate(block):
            state[j] = (state[j] + value) % P
        state = permute(state)
        state = [_to_mpz(v) for v in state]
        idx += 3
    return int(state[0])


def _self_check() -> None:
    """Re-derive every bundled KAT vector; abort if any disagree."""
    for kat in _raw["kat"]:
        got = permute([int(x, 16) for x in kat["input"]])
        expected = [int(x, 16) for x in kat["output"]]
        if got != expected:
            raise RuntimeError(
                f"Poseidon2 KAT mismatch:\n  input={kat['input']}\n  "
                f"got={got}\n  expected={expected}"
            )


# Fail fast at import time if our permutation diverges from gnark-crypto.
_self_check()
